File: src/agents/shared/health.py
"""
Health Check Module for Trading Agents

Provides standardized health check functionality for all AI agents
following the system-wide health check specification.
"""

import asyncio
import logging
import os
import time
from typing import Dict, Any, Optional
from datetime import datetime, timezone
from uuid import uuid4

import psutil
from fastapi import FastAPI, Request, HTTPException
from fastapi.responses import JSONResponse
from pydantic import BaseModel, Field

# from .database import DatabaseManager  # TODO: Implement when database module is ready
# from .kafka_client import KafkaManager  # TODO: Implement when Kafka module is ready

logger = logging.getLogger(__name__)


class HealthCheckResponse(BaseModel):
    """Standardized health check response model"""
    
    data: Dict[str, Any]
    error: Optional[str] = None
    correlation_id: str = Field(..., alias="correlation_id")


class HealthStatus(BaseModel):
    """Health status data model"""
    
    status: str  # healthy, degraded, unhealthy
    timestamp: str
    service: str
    version: str
    uptime: float
    environment: str
    response_time: int
    correlation_id: str
    checks: Dict[str, Dict[str, Any]]
    metadata: Dict[str, Any]


class HealthChecker:
    """Health check implementation for trading agents"""
    
    def __init__(
        self, 
        service_name: str, 
        version: str = "0.1.0",
        db_manager: Optional[Any] = None,  # DatabaseManager when implemented
        kafka_manager: Optional[Any] = None  # KafkaManager when implemented
    ):
        self.service_name = service_name
        self.version = version
        self.db_manager = db_manager
        self.kafka_manager = kafka_manager
        self.start_time = time.time()
    
    async def check_health(self, correlation_id: Optional[str] = None) -> HealthStatus:
        """
        Perform comprehensive health check
        
        Args:
            correlation_id: Request correlation ID for tracing
            
        Returns:
            HealthStatus: Complete health status information
        """
        start_time = time.time()
        correlation_id = correlation_id or str(uuid4())
        
        try:
            # Run all health checks concurrently
            check_tasks = [
                self._check_database(),
                self._check_kafka(),
                self._check_memory(),
                self._check_disk(),
                self._check_cpu(),
            ]
            
            results = await asyncio.gather(*check_tasks, return_exceptions=True)
            
            checks = {
                'database': self._process_result(results[0]),
                'kafka': self._process_result(results[1]),
                'memory': self._process_result(results[2]),
                'disk': self._process_result(results[3]),
                'cpu': self._process_result(results[4]),
            }
            
            # Determine overall status
            status = self._determine_overall_status(checks)
            
            health_status = HealthStatus(
                status=status,
                timestamp=datetime.now(timezone.utc).isoformat(),
                service=self.service_name,
                version=self.version,
                uptime=time.time() - self.start_time,
                environment=self._get_environment(),
                response_time=int((time.time() - start_time) * 1000),
                correlation_id=correlation_id,
                checks=checks,
                metadata=self._get_system_metadata()
            )
            
            return health_status
            
        except Exception as e:
            logger.exception(f"Health check failed: {e}")
            raise HTTPException(
                status_code=503, 
                detail=f"Health check failed: {str(e)}"
            )
    
    async def _check_database(self) -> Dict[str, Any]:
        """Check database connectivity and performance"""
        start_time = time.time()
        
        try:
            if not self.db_manager:
                return {
                    'status': 'skipped',
                    'message': 'Database manager not configured',
                    'response_time': int((time.time() - start_time) * 1000)
                }
            
            # Test database connection with a simple query  
            # TODO: Implement actual database connection test
            # await self.db_manager.execute_query("SELECT 1")
            await asyncio.sleep(0.01)  # Simulate database check
            
            return {
                'status': 'passed',
                'message': 'Database connection healthy',
                'response_time': int((time.time() - start_time) * 1000)
            }
            
        except Exception as e:
            return {
                'status': 'failed',
                'message': f'Database check failed: {str(e)}',
                'response_time': int((time.time() - start_time) * 1000)
            }
    
    async def _check_kafka(self) -> Dict[str, Any]:
        """Check Kafka connectivity and producer health"""
        start_time = time.time()
        
        try:
            if not self.kafka_manager:
                return {
                    'status': 'skipped',
                    'message': 'Kafka manager not configured',
                    'response_time': int((time.time() - start_time) * 1000)
                }
            
            # Test Kafka connectivity
            # TODO: Implement actual Kafka connectivity test
            # is_healthy = await self.kafka_manager.health_check()
            await asyncio.sleep(0.005)  # Simulate Kafka check
            is_healthy = True  # Simulated for now
            
            if is_healthy:
                return {
                    'status': 'passed',
                    'message': 'Kafka connection healthy',
                    'response_time': int((time.time() - start_time) * 1000)
                }
            else:
                return {
                    'status': 'failed',
                    'message': 'Kafka connection failed',
                    'response_time': int((time.time() - start_time) * 1000)
                }
                
        except Exception as e:
            return {
                'status': 'failed',
                'message': f'Kafka check failed: {str(e)}',
                'response_time': int((time.time() - start_time) * 1000)
            }
    
    async def _check_memory(self) -> Dict[str, Any]:
        """Check memory usage"""
        start_time = time.time()
        
        try:
            memory = psutil.virtual_memory()
            usage_percent = memory.percent
            
            if usage_percent > 90:
                status = 'failed'
                message = f'High memory usage: {usage_percent:.1f}%'
            elif usage_percent > 80:
                status = 'warning'
                message = f'Elevated memory usage: {usage_percent:.1f}%'
            else:
                status = 'passed'
                message = f'Memory usage normal: {usage_percent:.1f}%'
            
            return {
                'status': status,
                'message': message,
                'response_time': int((time.time() - start_time) * 1000),
                'details': {
                    'total': memory.total,
                    'available': memory.available,
                    'used': memory.used,
                    'percentage': usage_percent
                }
            }
            
        except Exception as e:
            return {
                'status': 'failed',
                'message': f'Memory check failed: {str(e)}',
                'response_time': int((time.time() - start_time) * 1000)
            }
    
    async def _check_disk(self) -> Dict[str, Any]:
        """Check disk usage"""
        start_time = time.time()
        
        try:
            disk = psutil.disk_usage('/')
            usage_percent = (disk.used / disk.total) * 100
            
            if usage_percent > 90:
                status = 'failed'
                message = f'High disk usage: {usage_percent:.1f}%'
            elif usage_percent > 80:
                status = 'warning'
                message = f'Elevated disk usage: {usage_percent:.1f}%'
            else:
                status = 'passed'
                message = f'Disk usage normal: {usage_percent:.1f}%'
            
            return {
                'status': status,
                'message': message,
                'response_time': int((time.time() - start_time) * 1000),
                'details': {
                    'total': disk.total,
                    'used': disk.used,
                    'free': disk.free,
                    'percentage': usage_percent
                }
            }
            
        except Exception as e:
            return {
                'status': 'failed',
                'message': f'Disk check failed: {str(e)}',
                'response_time': int((time.time() - start_time) * 1000)
            }
    
    async def _check_cpu(self) -> Dict[str, Any]:
        """Check CPU usage"""
        start_time = time.time()
        
        try:
            # Get CPU usage over a short interval
            cpu_percent = psutil.cpu_percent(interval=0.1)
            
            if cpu_percent > 90:
                status = 'failed'
                message = f'High CPU usage: {cpu_percent:.1f}%'
            elif cpu_percent > 80:
                status = 'warning'
                message = f'Elevated CPU usage: {cpu_percent:.1f}%'
            else:
                status = 'passed'
                message = f'CPU usage normal: {cpu_percent:.1f}%'
            
            return {
                'status': status,
                'message': message,
                'response_time': int((time.time() - start_time) * 1000),
                'details': {
                    'percentage': cpu_percent,
                    'count': psutil.cpu_count(),
                    'load_average': psutil.getloadavg() if hasattr(psutil, 'getloadavg') else None
                }
            }
            
        except Exception as e:
            return {
                'status': 'failed',
                'message': f'CPU check failed: {str(e)}',
                'response_time': int((time.time() - start_time) * 1000)
            }
    
    def _process_result(self, result: Any) -> Dict[str, Any]:
        """Process health check result, handling exceptions"""
        if isinstance(result, Exception):
            return {
                'status': 'failed',
                'message': f'Check failed with exception: {str(result)}',
                'response_time': 0
            }
        return result
    
    def _determine_overall_status(self, checks: Dict[str, Dict[str, Any]]) -> str:
        """Determine overall health status from individual checks"""
        failed_checks = [
            name for name, check in checks.items() 
            if check.get('status') == 'failed'
        ]
        
        warning_checks = [
            name for name, check in checks.items() 
            if check.get('status') == 'warning'
        ]
        
        if failed_checks:
            return 'unhealthy'
        elif warning_checks:
            return 'degraded'
        else:
            return 'healthy'
    
    def _get_environment(self) -> str:
        """Get current environment"""
        import os
        return os.getenv('ENVIRONMENT', 'development')
    
    def _get_system_metadata(self) -> Dict[str, Any]:
        """Get system metadata for diagnostics"""
        import platform
        import sys
        
        return {
            'python_version': sys.version,
            'platform': platform.platform(),
            'hostname': platform.node(),
            'pid': os.getpid(),
            'thread_count': psutil.Process().num_threads(),
        }


def setup_health_endpoint(app: FastAPI, health_checker: HealthChecker):
    """
    Setup health check endpoint on FastAPI application
    
    Args:
        app: FastAPI application instance
        health_checker: Health checker instance
    """
    
    @app.get("/health", response_model=HealthCheckResponse)
    async def health_check(request: Request):
        """Health check endpoint"""
        correlation_id = request.headers.get('x-correlation-id')
        
        try:
            health_status = await health_checker.check_health(correlation_id)
            
            # Determine HTTP status code based on health status
            status_code = 200
            if health_status.status == 'unhealthy':
                status_code = 503
            elif health_status.status == 'degraded':
                status_code = 200  # Still accepting traffic but with warnings
            
            return JSONResponse(
                content={
                    "data": health_status.dict(),
                    "error": None,
                    "correlation_id": health_status.correlation_id
                },
                status_code=status_code,
                headers={
                    "X-Correlation-ID": health_status.correlation_id,
                    "Cache-Control": "no-cache, no-store, must-revalidate"
                }
            )
            
        except HTTPException:
            raise
        except Exception as e:
            logger.exception("Unexpected error in health check")
            correlation_id = correlation_id or str(uuid4())
            
            return JSONResponse(
                content={
                    "data": {
                        "status": "unhealthy",
                        "timestamp": datetime.now(timezone.utc).isoformat(),
                        "service": health_checker.service_name,
                        "error": str(e),
                        "correlation_id": correlation_id
                    },
                    "error": str(e),
                    "correlation_id": correlation_id
                },
                status_code=503,
                headers={
                    "X-Correlation-ID": correlation_id
                }
            )

File: src/execution-engine/src/api/health.rs
/*!
 * Health Check Module for Execution Engine
 *
 * Provides standardized health check functionality for the high-performance
 * execution engine, following system-wide health check specifications.
 */

use std::collections::HashMap;
use std::time::{SystemTime, UNIX_EPOCH, Instant};
use serde::{Deserialize, Serialize};
use uuid::Uuid;

/// Standardized health check response
#[derive(Serialize, Deserialize, Debug)]
pub struct HealthResponse {
    pub data: HealthStatus,
    pub error: Option<String>,
    pub correlation_id: String,
}

/// Health status data structure
#[derive(Serialize, Deserialize, Debug)]
pub struct HealthStatus {
    pub status: String, // "healthy", "degraded", "unhealthy"
    pub timestamp: String,
    pub service: String,
    pub version: String,
    pub uptime: u64,
    pub environment: String,
    pub response_time: u64,
    pub correlation_id: String,
    pub checks: HashMap<String, CheckResult>,
    pub metadata: SystemMetadata,
}

/// Individual check result
#[derive(Serialize, Deserialize, Debug)]
pub struct CheckResult {
    pub status: String, // "passed", "failed", "warning", "skipped"
    pub message: String,
    pub response_time: u64,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub details: Option<serde_json::Value>,
}

/// System metadata for diagnostics
#[derive(Serialize, Deserialize, Debug)]
pub struct SystemMetadata {
    pub rust_version: String,
    pub platform: String,
    pub hostname: String,
    pub pid: u32,
    pub memory_usage: MemoryInfo,
    pub cpu_count: usize,
}

/// Memory information
#[derive(Serialize, Deserialize, Debug)]
pub struct MemoryInfo {
    pub resident_size: u64,
    pub virtual_size: u64,
}

pub struct HealthChecker {
    service_name: String,
    version: String,
    start_time: SystemTime,
}

impl HealthChecker {
    /// Create a new health checker instance
    pub fn new(service_name: String, version: String) -> Self {
        Self {
            service_name,
            version,
            start_time: SystemTime::now(),
        }
    }

    /// Perform comprehensive health check
    pub async fn check_health(&self, correlation_id: Option<String>) -> Result<HealthResponse, Box<dyn std::error::Error>> {
        let start_time = Instant::now();
        let correlation_id = correlation_id.unwrap_or_else(|| Uuid::new_v4().to_string());

        let mut checks = HashMap::new();

        // Run all health checks
        checks.insert("database".to_string(), self.check_database().await);
        checks.insert("kafka".to_string(), self.check_kafka().await);
        checks.insert("memory".to_string(), self.check_memory().await);
        checks.insert("disk".to_string(), self.check_disk().await);
        checks.insert("cpu".to_string(), self.check_cpu().await);
        checks.insert("mt_bridges".to_string(), self.check_mt_bridges().await);

        // Determine overall status
        let overall_status = self.determine_overall_status(&checks);

        let health_status = HealthStatus {
            status: overall_status,
            timestamp: self.current_timestamp(),
            service: self.service_name.clone(),
            version: self.version.clone(),
            uptime: self.uptime_seconds(),
            environment: self.get_environment(),
            response_time: start_time.elapsed().as_millis() as u64,
            correlation_id: correlation_id.clone(),
            checks,
            metadata: self.get_system_metadata(),
        };

        Ok(HealthResponse {
            data: health_status,
            error: None,
            correlation_id,
        })
    }

    /// Check database connectivity
    async fn check_database(&self) -> CheckResult {
        let start_time = Instant::now();

        // In a real implementation, this would test the database connection
        // For now, we'll simulate based on configuration
        match std::env::var("DATABASE_URL") {
            Ok(_) => {
                // Simulate database ping
                tokio::time::sleep(tokio::time::Duration::from_millis(10)).await;
                
                CheckResult {
                    status: "passed".to_string(),
                    message: "Database connection healthy".to_string(),
                    response_time: start_time.elapsed().as_millis() as u64,
                    details: None,
                }
            }
            Err(_) => CheckResult {
                status: "failed".to_string(),
                message: "DATABASE_URL not configured".to_string(),
                response_time: start_time.elapsed().as_millis() as u64,
                details: None,
            },
        }
    }

    /// Check Kafka connectivity
    async fn check_kafka(&self) -> CheckResult {
        let start_time = Instant::now();

        match std::env::var("KAFKA_BROKERS") {
            Ok(_) => {
                // Simulate Kafka health check
                tokio::time::sleep(tokio::time::Duration::from_millis(5)).await;
                
                CheckResult {
                    status: "passed".to_string(),
                    message: "Kafka connection healthy".to_string(),
                    response_time: start_time.elapsed().as_millis() as u64,
                    details: None,
                }
            }
            Err(_) => CheckResult {
                status: "failed".to_string(),
                message: "KAFKA_BROKERS not configured".to_string(),
                response_time: start_time.elapsed().as_millis() as u64,
                details: None,
            },
        }
    }

    /// Check memory usage
    async fn check_memory(&self) -> CheckResult {
        let start_time = Instant::now();

        // Get memory information using system calls or libraries
        let memory_info = self.get_memory_info();
        
        // For now, we'll use a simple heuristic
        // In production, you'd want to use proper system memory monitoring
        let status = if memory_info.resident_size > 1_000_000_000 { // > 1GB
            "warning"
        } else {
            "passed"
        };

        CheckResult {
            status: status.to_string(),
            message: format!("Memory usage: {:.2} MB", memory_info.resident_size as f64 / 1_000_000.0),
            response_time: start_time.elapsed().as_millis() as u64,
            details: Some(serde_json::to_value(&memory_info).unwrap()),
        }
    }

    /// Check disk space
    async fn check_disk(&self) -> CheckResult {
        let start_time = Instant::now();

        // In a real implementation, this would check actual disk usage
        // For containers, this is less critical
        CheckResult {
            status: "passed".to_string(),
            message: "Disk space adequate".to_string(),
            response_time: start_time.elapsed().as_millis() as u64,
            details: None,
        }
    }

    /// Check CPU usage
    async fn check_cpu(&self) -> CheckResult {
        let start_time = Instant::now();

        // Simple CPU check - in production you'd want proper CPU monitoring
        let cpu_count = num_cpus::get();
        
        CheckResult {
            status: "passed".to_string(),
            message: format!("CPU normal, {} cores available", cpu_count),
            response_time: start_time.elapsed().as_millis() as u64,
            details: Some(serde_json::json!({
                "cpu_count": cpu_count,
                "load_average": "not_implemented"
            })),
        }
    }

    /// Check MetaTrader bridge health
    async fn check_mt_bridges(&self) -> CheckResult {
        let start_time = Instant::now();

        // Check MT4 and MT5 bridge configurations
        let mt4_configured = std::env::var("MT4_SERVER").is_ok();
        let mt5_configured = std::env::var("MT5_SERVER").is_ok();

        let status = if mt4_configured || mt5_configured {
            "passed"
        } else {
            "warning"
        };

        let message = match (mt4_configured, mt5_configured) {
            (true, true) => "MT4 and MT5 bridges configured",
            (true, false) => "MT4 bridge configured, MT5 not configured",
            (false, true) => "MT5 bridge configured, MT4 not configured",
            (false, false) => "No MetaTrader bridges configured",
        };

        CheckResult {
            status: status.to_string(),
            message: message.to_string(),
            response_time: start_time.elapsed().as_millis() as u64,
            details: Some(serde_json::json!({
                "mt4_configured": mt4_configured,
                "mt5_configured": mt5_configured
            })),
        }
    }

    /// Determine overall health status from individual checks
    fn determine_overall_status(&self, checks: &HashMap<String, CheckResult>) -> String {
        let failed_count = checks.values().filter(|c| c.status == "failed").count();
        let warning_count = checks.values().filter(|c| c.status == "warning").count();

        if failed_count > 0 {
            "unhealthy".to_string()
        } else if warning_count > 0 {
            "degraded".to_string()
        } else {
            "healthy".to_string()
        }
    }

    /// Get current timestamp in ISO 8601 format
    fn current_timestamp(&self) -> String {
        let now = SystemTime::now();
        let datetime = chrono::DateTime::<chrono::Utc>::from(now);
        datetime.to_rfc3339()
    }

    /// Get uptime in seconds
    fn uptime_seconds(&self) -> u64 {
        self.start_time
            .elapsed()
            .unwrap_or_default()
            .as_secs()
    }

    /// Get environment name
    fn get_environment(&self) -> String {
        std::env::var("ENVIRONMENT").unwrap_or_else(|_| "development".to_string())
    }

    /// Get system metadata
    fn get_system_metadata(&self) -> SystemMetadata {
        SystemMetadata {
            rust_version: env!("CARGO_PKG_RUST_VERSION").to_string(),
            platform: std::env::consts::OS.to_string(),
            hostname: gethostname::gethostname().to_string_lossy().to_string(),
            pid: std::process::id(),
            memory_usage: self.get_memory_info(),
            cpu_count: num_cpus::get(),
        }
    }

    /// Get memory information
    fn get_memory_info(&self) -> MemoryInfo {
        // In a real implementation, you'd use proper system calls or libraries
        // This is a placeholder implementation
        MemoryInfo {
            resident_size: 50_000_000, // 50MB placeholder
            virtual_size: 100_000_000,  // 100MB placeholder
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[tokio::test]
    async fn test_health_check_creation() {
        let checker = HealthChecker::new(
            "execution-engine".to_string(),
            "0.1.0".to_string(),
        );

        let result = checker.check_health(None).await;
        assert!(result.is_ok());

        let response = result.unwrap();
        assert_eq!(response.data.service, "execution-engine");
        assert_eq!(response.data.version, "0.1.0");
        assert!(!response.data.correlation_id.is_empty());
    }

    #[tokio::test]
    async fn test_health_status_determination() {
        let checker = HealthChecker::new(
            "test-service".to_string(),
            "0.1.0".to_string(),
        );

        let mut checks = HashMap::new();
        checks.insert("test1".to_string(), CheckResult {
            status: "passed".to_string(),
            message: "OK".to_string(),
            response_time: 10,
            details: None,
        });
        checks.insert("test2".to_string(), CheckResult {
            status: "failed".to_string(),
            message: "Failed".to_string(),
            response_time: 20,
            details: None,
        });

        let status = checker.determine_overall_status(&checks);
        assert_eq!(status, "unhealthy");
    }
}